sbncode/CAFMaker/FillTrue.cxx

```diff
@@ -972,27 +994,40 @@ float ContainedLength(const TVector3 &v0, const TVector3 &v1,
 }//ContainedLength
 
 //------------------------------------------------
-caf::SRTrackTruth MatchTrack2Truth(const detinfo::DetectorClocksData &clockData, const std::vector<caf::SRTrueParticle> &particles, const std::vector<art::Ptr<recob::Hit>> &hits) {
+caf::SRTrackTruth MatchTrack2Truth(const detinfo::DetectorClocksData &clockData, const std::vector<caf::SRTrueParticle> &particles, const std::vector<art::Ptr<recob::Hit>> &hits,
+				   const std::map<int, caf::HitsEnergy> &all_hits_map) {
+
+  art::ServiceHandle<cheat::BackTrackerService> bt_serv;
 
   // this id is the same as the mcparticle ID as long as we got it from geant4
   std::vector<std::pair<int, float>> matches = CAFRecoUtils::AllTrueParticleIDEnergyMatches(clockData, hits, true);
   float total_energy = CAFRecoUtils::TotalHitEnergy(clockData, hits);
+  std::map<int, caf::HitsEnergy> track_hits_map = caf::SetupIDHitEnergyMap(hits, clockData, *bt_serv.get());
 
   caf::SRTrackTruth ret;
 
   ret.total_deposited_energy = total_energy / 1000. /* MeV -> GeV */;
 
   // setup the matches
   for (auto const &pair: matches) {
-    caf::ParticleMatch match;
+    caf::SRParticleMatch match;
     match.G4ID = pair.first;
     match.energy = pair.second / 1000. /* MeV -> GeV */;
+
+    caf::HitsEnergy track_matched_hits = track_hits_map.find(match.G4ID)->second;
+    caf::HitsEnergy all_matched_hits = all_hits_map.find(match.G4ID)->second;
+
+    match.hit_purity = (hits.size() != 0) ? track_matched_hits.nHits / (float) hits.size() : 0.;
+    match.energy_purity = (ret.total_deposited_energy > std::numeric_limits<float>::epsilon()) ? match.energy / ret.total_deposited_energy : 0.;
+    match.hit_completeness = (all_matched_hits.nHits != 0) ? track_matched_hits.nHits / (float) all_matched_hits.nHits : 0.;
+    match.energy_completeness = (all_matched_hits.totE > std::numeric_limits<float>::epsilon()) ? pair.second / all_matched_hits.totE : 0.;
```

Comment:
Are these tests against epsilon just due to paranoia about floating point comparisons? Test against zero is safe.
```suggestion
    match.hit_purity = (hits.size() != 0) ? track_matched_hits.nHits / (float) hits.size() : 0.;
    match.energy_purity = (ret.total_deposited_energy > 0) ? match.energy / ret.total_deposited_energy : 0.;
    match.hit_completeness = (all_matched_hits.nHits != 0) ? track_matched_hits.nHits / (float) all_matched_hits.nHits : 0.;
    match.energy_completeness = (all_matched_hits.totE > 0) ? pair.second / all_matched_hits.totE : 0.;
```

Reply:
Yeah they are a little paranoid when you put it like that!